cpp/oneapi/dal/exception.hpp

```diff
@@ -0,0 +1,103 @@
+/*******************************************************************************
+* Copyright 2020 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#pragma once
+
+#include <cstring>
+#include <memory>
+#include <new>
+#include <stdexcept>
+#include <system_error>
+
+namespace oneapi::dal {
+
+class exception {
```

Comment:
I believe, we must define virtual desctructor here. Try to build and run the following example, to understand why it's required.

```c++
#include <iostream>
#include <stdexcept>
#include <memory>

class exception {
public:
    ~exception() { std::cout << "exception killed" << std::endl; }
    virtual const char* what() const noexcept = 0;
};

class logic_error : public exception {
public:
    ~logic_error() { std::cout << "logic_error killed" << std::endl; }
};

class invalid_argument : public logic_error, public std::invalid_argument {
public:
    ~invalid_argument() { std::cout << "invalid_argument killed" << std::endl; }
    invalid_argument(const char* what_arg) : std::invalid_argument(what_arg) {}
    const char* what() const noexcept override;
};

int main(int argc, char const *argv[]) {
    exception *ia = new invalid_argument{"ddddd"};
    delete ia;

    return 0;
}
```

I know, we unlikely have exception allocation in the heap, but in any way, when we are designing a class, we should quarantine proper resource deallocation.

Reply:
```c++
class exception {
public:
    virtual ~exception() = default;
    virtual const char* what() const noexcept = 0;
};
```